### DISCUSSION TITLE
I can't get lidar to work with lds02rr_esp32.ino on ESP32 WROWER

### DISCUSSION THREAD

**Discussion by ArtemVolkov2103** [makerspet/support#9] [Support Q&A] [answered]
Hi! Today I received the LDS02RR lidar, and I came across the fact that it does not give any values. I've commented out the line `Serial.println(pwm_value);` and in Serial I get the output 
```
1023
2047
```
 When I commented out the lines 
```
//Serial.println("LDS motor error");
//xv_lds.enableMotor(false);
```
then the lidar does not spin at all, constantly displaying the message "LDS motor error" in Serial. I only changed the 
`#define LDS_MOTOR_EN_PIN      4 // LDS enable pin` because it's a free pin now. Maybe I missed something? I will be grateful if you know the answer.

**Comment by kaiaai** [comment]
@ArtemVolkov2103 I assume you mean "uncommented" those lines.

About uncommenting `Serial.println(pwm_value);` and getting `1023` and `2047` prints.

I believe value `1023` comes from LiDAR motor PID getting initialized at 50% PWM: `pwm_val = 0.5;` in `XV::XV()` xv_lib.h/cpp.

Then, `void xv_motor_pwm_callback(float pwm)` in `lds02rr_esp32.ino` gets called with value `pwm=0.5`. This float value of `0.5` gets converted to integer `1023` in this line `int pwm_value = ((1<<LDS_MOTOR_PWM_BITS)-1)*pwm;` Given that `#define LDS_MOTOR_PWM_BITS    11`, `((1<<11)-1)*0.5` calculates to `1023`.

The next PID update, I assume, increases `pwm` from `0.5` to `1.0` - because the motor takes time to spin up. That brings the integer `pwm` value up to `2047`.

About why uncommenting `//xv_lds.enableMotor(false);` stops the motor:
```
void loop() {
  ...
  if (!xv_lds.loop()) {
    // LDS motor error
    Serial.println("LDS motor error");
    xv_lds.enableMotor(false);
  }
}
```
Arduino loop() calls xv_lds.loop() in xv_lib.h/cpp
```
bool XV::loop() {
  ... 
  return motorCheck();
}
```
XV::loop() calls XV::motorCheck()
```
bool XV::motorCheck() {  // Make sure the motor RPMs are good else shut it down
  ...
  if (motor_enabled && ((scan_rpm < scan_rpm_min) || (scan_rpm > scan_rpm_max))) {
    scan_rpm_err++;
  } else {
    scan_rpm_err = 0;
  }
  ...
  return (scan_rpm_err > scan_rpm_err_thresh);
}
```
XV::motorCheck() checks LiDAR RPM every `motor_check_interval = 200;` msec. If the RPM is too high or too low, motorCheck() increments the `scan_rpm_err` RPM error count. Once the RPM error count exceeds `scan_rpm_err_thresh = 10;`, motorCheck() returns `false`. That triggers the `xv_lds.enableMotor(false);` call in the Arduino loop. This call shuts down the motor.

When you power up LDS02RR, its RPM starts from zero and starts increasing as the motor spins up faster and faster. It takes a few seconds or so for the RPM to get close to its target speed. Until then, `scan_rpm_err` keeps getting incremented, reaches its threshold `scan_rpm_err_thresh = 10;`, causes `xv_lib.loop()` to error out and triggers `xv_lds.enableMotor(false);` the motor shutdown. The motor shutdown causes the RPM to decrease, which keeps the error count incrementing indefinitely.

So the LiDAR simply does not have enough time to spin up to its target speed. If I remember correctly, I had the same issue. You can increase `scan_rpm_err_thresh = 10;` to get around this error.

XV::loop() lines are not my code. That routine came from https://github.com/getSurreal/XV_Lidar_Controller Looking at this comment `scan_rpm_err_thresh = 10;  // 2 seconds (10 * 200ms) to shutdown motor with improper RPM and high voltage` I guess this error mechanism is supposed to turn the LiDAR off if the motor voltage is too high or too low.

---

When  you start your bring up, first off, make sure LDS02RR serial communication works with ESP32. Check that your `Serial1.read()` returns any data at all. If there is no data, check TX, RX pin assignments, wire connections. Check that LDS02RR is powered up correctly (5V, not 3.3V). Check LDS02RR TX with an oscilloscope, if you have one. Make sure your Arduino esp32 toolchain version is NOT 2.0.15. I assume the baud rate is already correct because it is hardcoded. Double-check that you indeed have the LDS02RR model - not LDS01RR.

LDS02RR sends distance values only when it spins close to its target RPM. In other words, you will not be getting distance values if the RPM is too fast or too slow. When the speed is too slow or too fast LDS02RR will be sending out short packets that contain its RPM value - so that the (external) PID/PWM controller can drive the LiDAR's motor to reach the correct RPM.

**Reply by ArtemVolkov2103** [reply]
Thanks for the detailed answer. 
I slightly changed the logic of `XV::motorCheck()`, and the lidar began to spin for 2 seconds, and then stop. Still, this is not what I needed.
 You said that when the RPM is too small or too large, the lidar should send the RPM. I've never seen Lidar send anything yet, even if i use 
``` 
while (Serial1.available() > 0) { // read byte from LDS
    Serial.print("LDS data: ");
    Serial.println(Serial1.read());
    xv_lds.processByte(Serial1.read());
  }
```
there is still nothing. No real RPM in Serial, always only zero by `Serial.print(xv_lds.getScanRPM());`

Judging by the compiler's output, the Arduino ESP32 toolchain version is 2.0.14
![image](https://github.com/makerspet/support/assets/82831826/6861d683-f612-476f-b2c2-2aeb51c44b90)

Does this mean that my lidar is broken?

**Reply by kaiaai** [reply]
@ArtemVolkov2103 have you probed your LiDAR and ESP32 with an oscilloscope, extensively? If not, it's too early to call it broken.

Try getting an oscilloscope, if you don't have one. Do you have a digital multimeter? This is what I would do.

Using a digital multimeter:
- probe ground for continuity between your ESP32 and the little PCB mounted on the LiDAR
- probe 5V for continuity between ESP32 and LiDAR
- measure power voltage at the LiDAR PCB, it should be 5V
- probe ESP32-to-LiDAR continuity for TX, RX. The LiDAR may have a resistor in series with TX (same for RX).

Using an oscilloscope:
- probe 5V on the LiDAR PCB
- probe TX, RX on the LiDAR PCB. TX should be toggling like a serial port output.
- probe TX, RX on the ESP32 - make sure ESP32 receives the TX signal

Also, feel free to post a clear photo of your wiring.

**Reply by ArtemVolkov2103** [reply, marked as answer]
Hello! It looks like I've solved the problem!
Thank you for advising me to use an oscilloscope. Unfortunately, I don't have an oscilloscope, but for this purpose, the ESP32 itself was enough for me. I found out that the LiDAR changes the state of the TX during rotation, which meant its minimum performance. 
Next, I read the ESP32 WROVER documentation again and remembered that 17 and 16 GPIOs cannot be used for this model, since they work with QSPI. I started looking for other pin combinations that the LiDAR and found pins 32 (TX) and 33 (RX). 
Now, the LiDAR is rotating at the desired speed and returning the measured data. 
I suggest adding the following information to the README.md: when using the ESP32 WROVER board, it's worth using pins 32 and 33 instead of pins 16 and 17. I connected the LiDAR to pin 33, and it works perfectly. 
I would like to thank you once again for your hard work!

**Reply by kaiaai** [reply]
@ArtemVolkov2103 great, thank you, I've updated the README.md
